### ISSUE TITLE
BulkInsert: MultiTenant->Tenant Per Database

### ISSUE THREAD

**Issue by kedwardsRenWeb** [zzzprojects/EntityFramework-Extensions#75]
So we have multiple webservers with multiple database servers. In an effort to optimize our ConnPools, our DB ConnectionString = DBServerX:Database1.  Immediately after opening our Connection - we changeDatabase to be the correct database.  This creates a situation where the SQLConnection's ConnectionString is 'stale'.  

My question is as follows:

Does BulkInsert use the ConnectionString or the actual Server and Database of the Connection? 

I can confirm from debugging that the Connection is correct, but I'm 99% sure that BulkInsert is trying to insert into a table based on the original connection string, vs the server and database which the SqlConnection is using.

Exception Trace (with some specific info removed)
2018-01-19 17:32:34,518 [74] DEBUG Server: [SERVER]Database: [DBNAME]
2018-01-19 17:32:35,926 [74] FATAL 
System.InvalidOperationException: Cannot access destination table '[dbo].[ENTITYNAME]'. ---> System.Data.SqlClient.SqlException: Invalid object name 'dbo.ENTITYNAME'.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.TdsParser.Run(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj)
   at System.Data.SqlClient.SqlBulkCopy.RunParser(BulkCopySimpleResultSet bulkCopyHandler)
   at System.Data.SqlClient.SqlBulkCopy.CreateAndExecuteInitialQueryAsync(BulkCopySimpleResultSet& result)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestAsync(CancellationToken cts, TaskCompletionSource`1 source)
   --- End of inner exception stack trace ---
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestAsync(CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalAsync(CancellationToken ctoken)
   at System.Data.SqlClient.SqlBulkCopy.WriteRowSourceToServerAsync(Int32 columnCount, CancellationToken ctoken)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServer(IDataReader reader)
   at EntityFramework.BulkInsert.Providers.EfSqlBulkInsertProviderWithMappedDataReader.Run[T](IEnumerable`1 entities, SqlTransaction transaction, BulkInsertOptions options) in c:\dev\EntityFramework.BulkInsert\dev\Src\EntityFramework.BulkInsert\Providers\EfSqlBulkInsertProviderWithMappedDataReader.cs:line 48
   at EntityFramework.BulkInsert.Providers.ProviderBase`2.Run[T](IEnumerable`1 entities, BulkInsertOptions options) in c:\dev\EntityFramework.BulkInsert\dev\Src\EntityFramework.BulkInsert\Providers\ProviderBase.cs:line 105
   at EntityFramework.BulkInsert.Extensions.BulkInsertExtension.BulkInsert[T](DbContext context, IEnumerable`1 entities, SqlBulkCopyOptions sqlBulkCopyOptions, Nullable`1 batchSize) in c:\dev\EntityFramework.BulkInsert\dev\Src\EntityFramework.BulkInsert\Extensions\BulkInsertExtension.cs:line 96
   at EntityFramework.BulkInsert.Extensions.BulkInsertExtension.BulkInsert[T](DbContext context, IEnumerable`1 entities, Nullable`1 batchSize) in c:\dev\EntityFramework.BulkInsert\dev\Src\EntityFramework.BulkInsert\Extensions\BulkInsertExtension.cs:line 76

**Comment by JonathanMagnan** [comment]
Hello @kedwardsRenWeb,

Yes, you are right.

It's changing database a common scenario in your application?

We have a method named `RefreshModel`

```csharp
context.RefreshModel();
```

It should works after you refresh the model.

Best Regards,

Jonathan

**Comment by kedwardsRenWeb** [comment]
It is because we have 22 web servers and 8000 client DBs srpead across 10 DB servers.

We blowup connection pools if we ha EA one connString Server/Database combo....


How exactly does refreshModel work (internally)?



Get Outlook for Android<https://aka.ms/ghei36>

**Comment by JonathanMagnan** [comment]
Hello @kedwardsRenWeb ,

This solution will not work four you then as you will change often the `connection` from what I understand.

In short, `RefreshModel` clean the cache everything related to the current context which has saved some information using the connection. In your case, that will be too long and this solution will not work for a concurrent scenario.

We currently investigate two solutions:
- Append by default the current database information to the context (You will only need to upgrade here)
- Keep the current cache key but add options to append current database information (You will only need to set an option to true when the application start)

In both case, the minimum of work will be required on your side.

We will try to implement the chosen solution later today.

Best Regards,

Jonathan

**Comment by kedwardsRenWeb** [comment]
To add clarity to our specific situation: 
our DbContext is instantiated **one time per request**.  We only call ChangeDatabase once per request...  so relative to 'all requests' we change it often... but relative to the scope of a single request: it only happens one time.

**Comment by kedwardsRenWeb** [comment]
I've found a workaround.  The issue isn't your code.
The issue was SqlBulkCopy.

My issue is essentially the fully qualified SQL table name isn't known until runtime.

I implemented my own extension method of BulkInsert that utilized SqlBulkCopy.  The one difference is that I pass in my DestinationTableName as a fullyQualifiedName:  {database}.{schema}.{tablename}

I used an extension method that would convert an IEnumerable into a DataTable.


Then I can pass in the fully qualified name and the DataTable into my BulkInsert method.
